text/0000-bee-network/0000-bee-network.md

```diff
@@ -0,0 +1,295 @@
+
+The functionalities of this crate are relatively low-level from a node perspective in the sense, that they are independent from any specifics defined by the IOTA protocol. Consequently, it has no dependencies on other crates of the framework, and can easily be reused in different contexts.
+
+The aim of this crate is to make it simple and straightforward for developers to build layers on top of it (e.g. a protocol layer) by abstracting away all the underlying networking logic. It is therefore a lot easier for them to focus on the modelling of other important aspects of a node software, like:
+
+* `Peer`s: other nodes in the network (other Bees and Hornets),
+* `Message`es: the information exchanged between peers, and its serialization/deserialization.
+
+Given some identifier `epid`, sending a message to its corresponding endpoint becomes a single line of asynchronous code:
+
+```rust
+network.send(SendMessage { epid, bytes: "hello".as_bytes() }).await?;
+```
+
+The purpose of this crate is to provide the following functionalities:
+* maintain a list of endpoints (not peers!),
+* establish and maintain connections with endpoints,
+* allow to send, multicast or broadcast byte encoded messages of variable size to any of the endpoints,
+* allow to receive byte encoded messages of variable size from any of the endpoints,
+* reject connections from unknown, i.e. not whitelisted, endpoints,
+* manage all connections and message transfers asynchronously,
+* provide an extensible, yet convenient-to-use, and documented API.
+
+The key design decisions are now being discussed in the following sub-sections.
+
+## async/await
+
+This crate is by nature very much dependent on events happening outside of the control of the program, e.g. listening for incoming connections from peers, waiting for packets on a specific socket, etc. Hence, this crate under the hood makes heavy use of Rust's concurrency abstractions. Luckily, with the stabilization of the `async/await` syntax, writing asynchronous code has become almost as easy to read, write, and maintain as synchronous code. In an experimental implementation of this crate the [`async_std`](https://github.com/async-rs/async-std) library was used, which comes with asynchronous drop-in replacements for their synchronous equivalents found in Rust's standard library `std`. Additionally, asynchronous mpsc channels were taken from the [`futures`](https://github.com/rust-lang/futures-rs) crate.
+
+## Message passing
+
+This crate favors message passing via channels over globally shared state and locks. Instead of keeping the list of endpoints in a globally accessible hashmap this crate separates and moves such state into workers that run asynchronously, and are listening for commands and events to act upon, and also notify listeners by sending events.
+
+## Initialization
+
+In order for this crate to be used, it has to be initialized first by providing a `NetworkConfig`:
+
+```rust
+struct NetworkConfig {
```

Comment:
Should `NetworkConfig` be required to implement `Default` for future compatibility?

Reply:
Can you elaborate ? According to RFC 31, `NetworkConfigBuilder` has to implement `Default`, then it's just a matter of finishing it ?

Reply:
@zesterer yes, please elaborate? @thibault-martinez  `NetworkConfig` is supposed to be created through the builder only, right?

Reply:
Ah, fair enough. All good then :)